scripts/jenkins/ardana/gerrit/build_test_package.py

```diff
@@ -434,50 +290,48 @@ def build_test_packages(change_ids, obs_linked_project, home_project,
     # 'gerrit_project': Package()
     packages = {}
 
-    # Grab each change for the supplied change_ids. As we go through the
-    # changes the change_ids list may expand with dependencies. These are also
-    # processed. If a change has already been processed we skip it to avoid
-    # circular dependencies.
-    try_branches = ['master']
+    # The target branch associated with the first change is used for
+    # all changes
+    branch = None
+
+    # Grab each change for the supplied change_ids
+    changes = []
```

Comment:
What we could do is create a Gerrit() class that simply holds a cache of changes in a dict. Then rather than reloading each GerritChange and having to do extra API queries we could pull them out of the Gerrit class. I know the dependency checks etc are cached already, but this would stop from double-loading a dependency that has been referred to by multiple supplied change_ids.

Then by doing this, we always have a 1-to-1 mapping of GerritChange objects to what we are tracking. Thus we could use a set() for changes and then when we doing a union join with any dependant changes we already weed out the duplicate or potential circular dependencies.

Reply:
Yep, fair enough. Absolutely something that can be done later. We'll probably find that while it's not perfectly optimised that gerrit can handle a decent load and it won't be necessary.